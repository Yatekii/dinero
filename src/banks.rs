mod ibkr;
mod neon;
mod ubs;

use std::path::Path;

use anyhow::{Context, Result};
use chrono::NaiveDate;
use ts_rs::TS;

use crate::cli::BankFormat;

pub fn parse(name: String, content: String, format: BankFormat) -> Result<Vec<LedgerRecord>> {
    Ok(match format {
        BankFormat::Neon => neon::Neon::parse(name, content),
        BankFormat::Ubs => ubs::Ubs::parse(name, content),
        BankFormat::Ibkr => ibkr::Ibkr::parse(name, content),
    }?
    .ledgers[0]
        .records
        .clone())
}

pub fn load(
    name: String,
    path: impl AsRef<Path>,
    format: BankFormat,
) -> anyhow::Result<Vec<LedgerRecord>> {
    match format {
        BankFormat::Neon => load_inner::<neon::Neon>(name, path),
        BankFormat::Ubs => load_inner::<ubs::Ubs>(name, path),
        BankFormat::Ibkr => load_inner::<ibkr::Ibkr>(name, path),
    }
}

fn load_inner<T: Parser>(
    name: String,
    path: impl AsRef<Path>,
) -> anyhow::Result<Vec<LedgerRecord>> {
    let mut records = vec![];
    for dir_entry in std::fs::read_dir(&path).with_context(|| {
        format!(
            "could not read ledger directory {}",
            path.as_ref().display()
        )
    })? {
        let dir_entry = dir_entry?;

        let path = dir_entry.path();
        let loaded = T::parse(
            name.clone(),
            std::fs::read_to_string(&path)
                .with_context(|| format!("could not read ledger CSV {}", path.display()))?,
        )?
        .ledgers[0]
            .records
            .clone();

        records.extend(loaded);
    }
    Ok(records)
}

pub trait Parser {
    fn parse(name: String, content: String) -> Result<ParsedAccount>;
}

#[derive(Debug)]
pub struct ParsedAccount {
    ledgers: Vec<ParsedLedger>,
}

#[derive(Debug)]
pub struct ParsedLedger {
    name: String,
    records: Vec<LedgerRecord>,
}

#[derive(Debug, Clone, PartialEq, serde::Deserialize)]
pub struct LedgerRecord {
    pub date: NaiveDate,
    pub amount: f64,
    pub description: String,
    pub category: String,
    pub symbol: String,
}

#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize, TS)]
pub struct ExtendedLedgerRecord {
    #[ts(type = "number")]
    pub date: NaiveDate,
    pub amount: f64,
    pub description: String,
    pub original_description: String,
    pub category: String,
    pub original_category: String,
    pub comments: String,
    pub checked: bool,
    pub symbol: String,
}
