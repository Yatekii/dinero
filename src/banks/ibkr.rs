use std::io::Cursor;

use anyhow::{bail, Error};
use chrono::NaiveDate;
use csv::{Reader, ReaderBuilder};

use super::{LedgerRecord, ParsedAccount, Parser};

pub struct Ibkr {}

impl Parser for Ibkr {
    fn parse(name: String, content: String) -> anyhow::Result<ParsedAccount> {
        let mut ledgers = vec![];
        let lines = content.lines().collect::<Vec<_>>();
        let mut end = lines.len();
        for (index, line) in lines.iter().enumerate().rev() {
            if line.starts_with("\"HEADER\"") {
                let content = &lines[index..end].join("\n");

                let reader = ReaderBuilder::new()
                    .delimiter(b',')
                    .from_reader(Cursor::new(&content));

                let records = if line.contains("\"CTRN\"") {
                    parse_cash_transactions(reader)?
                } else if line.contains("\"TRNT\"") {
                    parse_stock_transactions(reader)?
                } else {
                    bail!("Unknown export")
                };

                ledgers.push(super::ParsedLedger {
                    name: name.clone(),
                    records,
                });

                end = index;
            }
        }

        Ok(ParsedAccount { ledgers })
    }
}

fn parse_stock_transactions(
    mut reader: Reader<Cursor<&&String>>,
) -> Result<Vec<LedgerRecord>, Error> {
    let records = reader
        .deserialize::<StockRecord>()
        .map(|v| {
            v.map(|v| LedgerRecord {
                date: v.date,
                amount: v.amount,
                description: v.description,
                category: "Broker".to_string(),
                symbol: v.symbol,
            })
        })
        .collect::<Result<_, _>>()?;

    Ok(records)
}

#[derive(Debug, serde::Deserialize)]
struct StockRecord {
    #[serde(rename = "TradeDate")]
    date: NaiveDate,
    #[serde(rename = "Quantity")]
    amount: f64,
    #[serde(rename = "Description")]
    description: String,
    #[serde(rename = "Symbol")]
    symbol: String,
}

fn parse_cash_transactions(
    mut reader: Reader<Cursor<&&String>>,
) -> Result<Vec<LedgerRecord>, Error> {
    let records = reader
        .deserialize::<CashRecord>()
        .map(|v| {
            v.map(|v| LedgerRecord {
                date: v.date,
                amount: v.amount,
                description: v.description,
                category: "Broker".to_string(),
                symbol: v.symbol,
            })
        })
        .collect::<Result<_, _>>()?;

    Ok(records)
}

#[derive(Debug, serde::Deserialize)]
struct CashRecord {
    #[serde(rename = "SettleDate")]
    date: NaiveDate,
    #[serde(rename = "Amount")]
    amount: f64,
    #[serde(rename = "Description")]
    description: String,
    #[serde(rename = "Symbol")]
    symbol: String,
}

#[cfg(test)]
mod tests {
    use crate::banks::Parser;

    const TRANSACTIONS: &str = r#"Trade date;Trade time;Booking date;Value date;Currency;Debit;Credit;Individual amount;Balance;Transaction no.;Description1;Description2;Description3;Footnotes;
2021-01-29;;2021-01-29;2021-01-31;CHF;-10.00;;;22113.15;BL01529HJ0125142;"Balance closing of service prices";;"Transaction no. BL01529HJ0125142";;
2021-01-27;;2021-01-28;2021-01-28;CHF;-503.50;;;22123.15;9906527KH9626550;"UBS Switzerland AG,c/o UBS Card Center";"VIS1W WIDERSPRUCH AN UBS INNERT 30 TAGEN, direct debit";"Reference no. 70 03130 00000 00326 20405 08188, Account no. IBAN: CH25 0023 0230 0129 5305 U, Costs: LSV direct debit, Transaction no. 9906527KH9626550";;
2021-01-27;;2021-01-27;2021-01-27;CHF;-62.75;;;22626.65;9906026TO5927425;"CAISSE DES MEDECINS,1211 GENEVE";"e-banking payment order";"Reference no. 05 62945 01846 09123 50099 10841, Account no. 01-037005-8, Costs: E-Banking domestic, Transaction no. 9906026TO5927425";;
2021-01-22;;2021-01-22;2021-01-22;CHF;;9.59;;22689.40;9999022ZC8003633;"STRIPE PAYMENTS UK LTD,9TH FLOOR, 107 CHEAPSIDE GB - LONDO, N EC2V 6DN";"credit";"Reason for payment: GITHUB SPONSORS R8N0A2, Costs: Incoming SIC-payment, Transaction no. 9999022ZC8003633";;
2021-01-22;;2021-01-22;2021-01-22;CHF;;3688.00;;22679.81;9999022ZC7962684;"TECHNOKRAT GMBH,UNTERROHRSTRASSE 5, 8952 SCHLIEREN, CH";"credit";"Reason for payment: LOHN, Costs: Incoming SIC-payment, Transaction no. 9999022ZC7962684";;
2021-01-18;;2021-01-18;2021-01-18;CHF;-482.80;;;18991.81;BF21018DJ2743561;"CSS KRANKEN-VERSICHERUNG,AG, LUZERN";"EBILL-RECHNUNG, PayNet Order";"Reference no. 00 00002 64029 44001 71051 49861, Account no. 01-070393-3, Costs: E-Banking domestic, Transaction no. BF21018DJ2743561";;
2021-01-13;;2021-01-13;2021-01-13;CHF;-820.00;;;19474.61;9906012TI0715972;"Stichting DEGIRO,NL";"e-banking payment order";"Reason for payment: 88391-243702-5FFD9A1F-8BB5, Account no. IBAN: CH82 0871 0039 1145 1200 2, Costs: E-Banking domestic, Transaction no. 9906012TI0715972";;
2021-01-13;;2021-01-13;2021-01-13;CHF;-20.00;;;20294.61;9906013GK1225794;"ROTH, JONAS, Debit UBS TWINT";;"Reason for payment: +41798181674, TWINT-Acc.:+41799607130, Transaction no. 9906013GK1225794";;
2021-01-04;;2021-01-04;2021-01-04;CHF;-350.00;;;20314.61;9906501KH9243834;"GENERALI PERSONENVERSICHERUNGEN AG,8134 ADLISWIL";"FORT1 WIDERSPRUCH AN UBS INNERT 30 TAGEN, direct debit";"Reference no. 91 34065 09797 16210 10157 11525, Costs: LSV direct debit, Transaction no. 9906501KH9243834";;
2021-01-04;;2021-01-04;2021-01-04;CHF;-281.35;;;20664.61;BH21004DJ2411724;"CSS KRANKEN-VERSICHERUNG,AG, LUZERN";"EBILL-RECHNUNG, PayNet Order";"Reference no. 00 00002 64029 44002 90851 13989, Account no. 01-070393-3, Costs: E-Banking domestic, Transaction no. BH21004DJ2411724";;"#;

    #[test]
    fn parse() {
        let _df = super::Ibkr::parse("IBKR".to_string(), TRANSACTIONS.into()).unwrap();
    }
}
