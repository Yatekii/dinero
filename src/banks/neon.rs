use std::io::Cursor;

use chrono::NaiveDate;
use csv::ReaderBuilder;

use super::{LedgerRecord, ParsedAccount, ParsedLedger, Parser};

pub struct Neon {}

impl Parser for Neon {
    fn parse(name: String, content: String) -> anyhow::Result<ParsedAccount> {
        let records = ReaderBuilder::new()
            .delimiter(b';')
            .from_reader(Cursor::new(&content))
            .deserialize::<Record>()
            .filter(|v| (v.as_ref()).map_or("", |v| &v.spaces) == "no")
            .map(|v| {
                v.map(|v| LedgerRecord {
                    date: v.date,
                    amount: v.amount,
                    description: v.description,
                    category: v.category,
                    symbol: "CHF".to_string(),
                })
            })
            .collect::<Result<_, _>>()?;

        Ok(ParsedAccount {
            ledgers: vec![ParsedLedger { name, records }],
        })
    }
}

#[derive(Debug, serde::Deserialize)]
struct Record {
    #[serde(rename = "Date")]
    date: NaiveDate,
    #[serde(rename = "Amount")]
    amount: f64,
    #[serde(rename = "Original amount")]
    _original_amount: Option<f64>,
    #[serde(rename = "Original currency")]
    _original_currency: Option<String>,
    #[serde(rename = "Exchange rate")]
    _exchange_rate: Option<f64>,
    #[serde(rename = "Description")]
    description: String,
    #[serde(rename = "Subject")]
    _subject: String,
    #[serde(rename = "Category")]
    category: String,
    #[serde(rename = "Tags")]
    _tags: String,
    #[serde(rename = "Wise")]
    _wise: String,
    #[serde(rename = "Spaces")]
    spaces: String,
}

#[cfg(test)]
pub mod tests {
    use crate::banks::Parser;

    use super::Neon;

    #[test]
    pub fn parse_export() {
        const TRANSACTIONS: &str = r#""Date";"Amount";"Original amount";"Original currency";"Exchange rate";"Description";"Subject";"Category";"Tags";"Wise";"Spaces"
"2019-09-17";"-1009.00";"";"";"";"Urech Optik";;"health";"";"no";"no"
"2019-08-02";"-200.00";"";"";"";"ZKB ZH HB SIHLQUAI 2";;"cash";"";"no";"no"
"2019-07-22";"-150.00";"";"";"";"Hanspeter Schoop";"ANZAHLUNG";"uncategorized";"";"no";"no"
"2019-07-22";"-439.70";"";"";"";"Generali";"000000002198972910070138387";"finances";"";"no";"no"
"2019-07-15";"-92.20";"";"";"";"Urbach Optik";"000000000000000041017313152";"health";"";"no";"no"
"2019-06-03";"-140.00";"";"";"";"Regionalpolizei Lenzburg";"000000000000000003478000113";"finances";"";"no";"no"
"2019-05-10";"30.00";"";"";"";"neon Switzerland AG";;"income";"";"no";"no"
"2019-05-10";"30.00";"";"";"";"neon Switzerland AG";;"income";"";"no";"no"
"2019-05-09";"3000.00";"";"";"";"Technokrat GmbH";"Lohn";"income_salary";"income";"no";"no"
"#;
        insta::assert_debug_snapshot!(
            Neon::parse("Neon".to_string(), TRANSACTIONS.to_string()).unwrap()
        );
    }
}
