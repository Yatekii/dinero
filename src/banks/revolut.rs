use std::io::Cursor;

use chrono::{
    format::{parse, Fixed, Item, Numeric, Pad, Parsed},
    NaiveDate,
};
use csv::ReaderBuilder;
use serde::{de::Visitor, Deserializer};

use super::{LedgerRecord, ParsedAccount, ParsedLedger, Parser};

pub struct Revolut {}

impl Parser for Revolut {
    fn parse(name: &str, content: String) -> anyhow::Result<ParsedAccount> {
        let records = ReaderBuilder::new()
            .delimiter(b',')
            .from_reader(Cursor::new(&content))
            .deserialize::<Record>()
            .filter(|v| (v.as_ref()).map_or("", |v| &v.state) == "COMPLETED")
            .filter_map(|v| {
                v.map(|v| {
                    v.date.map(|date| LedgerRecord {
                        date,
                        amount: v.amount,
                        description: v.description,
                        category: v.category,
                        symbol: v.currency,
                    })
                })
                .transpose()
            })
            .collect::<Result<_, _>>()?;

        Ok(ParsedAccount {
            ledgers: vec![ParsedLedger {
                name: name.to_string(),
                records,
            }],
        })
    }
}

#[derive(Debug, serde::Deserialize)]
struct Record {
    #[serde(rename = "Completed Date", deserialize_with = "parse_date_with_time")]
    date: Option<NaiveDate>,
    #[serde(rename = "Amount")]
    amount: f64,
    #[serde(rename = "Description")]
    description: String,
    #[serde(rename = "Type")]
    category: String,
    #[serde(rename = "Currency")]
    currency: String,
    #[serde(rename = "State")]
    state: String,
}

fn parse_date_with_time<'de, D: Deserializer<'de>>(de: D) -> Result<Option<NaiveDate>, D::Error> {
    const ITEMS: &[Item<'static>] = &[
        Item::Numeric(Numeric::Year, Pad::Zero),
        Item::Space(""),
        Item::Literal("-"),
        Item::Numeric(Numeric::Month, Pad::Zero),
        Item::Space(""),
        Item::Literal("-"),
        Item::Numeric(Numeric::Day, Pad::Zero),
        Item::Space(""),
        Item::Numeric(Numeric::Hour, Pad::Zero),
        Item::Space(""),
        Item::Literal(":"),
        Item::Numeric(Numeric::Minute, Pad::Zero),
        Item::Space(""),
        Item::Literal(":"),
        Item::Numeric(Numeric::Second, Pad::Zero),
        Item::Fixed(Fixed::Nanosecond),
        Item::Space(""),
    ];

    struct DateVisitor;
    struct MaybeDateVisitor;

    impl Visitor<'_> for DateVisitor {
        type Value = NaiveDate;

        fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
        where
            E: serde::de::Error,
        {
            let mut parsed = Parsed::new();
            parse(&mut parsed, v, ITEMS.iter()).map_err(E::custom)?;
            parsed.to_naive_date().map_err(E::custom)
        }

        fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
            formatter.write_str("expected a date string of the format YYYY-MM-DD hh:mm::ss")
        }
    }

    impl<'de> Visitor<'de> for MaybeDateVisitor {
        type Value = Option<NaiveDate>;

        fn visit_none<E>(self) -> Result<Self::Value, E>
        where
            E: serde::de::Error,
        {
            Ok(None)
        }

        fn visit_some<D: Deserializer<'de>>(self, de: D) -> Result<Self::Value, D::Error> {
            Ok(Some(de.deserialize_str(DateVisitor)?))
        }

        fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
            formatter
                .write_str("expected a date string of the format YYYY-MM-DD hh:mm::ss or no value")
        }
    }

    de.deserialize_option(MaybeDateVisitor)
}

#[cfg(test)]
pub mod tests {
    use crate::banks::Parser;

    use super::Revolut;

    #[test]
    pub fn parse_export() {
        const TRANSACTIONS: &str = r#"Type,Product,Started Date,Completed Date,Description,Amount,Fee,Currency,State,Balance
CARD_PAYMENT,Current,2023-01-02 03:54:18,2023-01-02 13:11:10,migrolino,-7.50,0.00,CHF,COMPLETED,91.65
TOPUP,Current,2023-01-05 09:14:08,2023-01-05 09:14:30,Auto Top-Up by *0382,75.00,0.00,CHF,COMPLETED,166.65
CARD_PAYMENT,Current,2023-01-05 09:14:00,2023-01-06 23:25:07,CCBill,-19.74,0.00,CHF,COMPLETED,146.91
CARD_PAYMENT,Current,2023-01-08 03:42:39,2023-01-08 11:34:11,GitHub,-7.44,0.07,CHF,COMPLETED,139.40
CARD_PAYMENT,Current,2023-01-12 13:55:31,2023-01-18 21:20:55,Contabo,-6.48,0.00,CHF,COMPLETED,132.92
CARD_PAYMENT,Current,2023-01-22 14:06:23,2023-01-28 22:57:06,Contabo,-11.32,0.11,CHF,COMPLETED,121.49
CARD_PAYMENT,Current,2023-02-04 09:14:03,2023-02-06 22:47:13,CCBill,-20.02,0.20,CHF,COMPLETED,101.27
CARD_PAYMENT,Current,2023-02-08 03:56:44,2023-02-08 11:32:47,GitHub,-18.14,0.00,CHF,COMPLETED,83.13
CARD_PAYMENT,Current,2023-02-12 13:52:03,2023-02-18 22:24:06,Contabo,-6.38,0.06,CHF,COMPLETED,76.69
TOPUP,Current,2023-02-22 13:54:55,2023-02-22 13:54:57,Auto Top-Up by *0382,75.00,0.00,CHF,COMPLETED,151.69
CARD_PAYMENT,Current,2023-02-22 13:54:54,2023-02-28 21:56:47,Contabo,-11.18,0.00,CHF,COMPLETED,140.51
CARD_PAYMENT,Current,2023-03-01 21:55:56,2023-03-02 11:41:17,Google Cloud,-7.38,0.00,CHF,COMPLETED,133.13
CARD_PAYMENT,Current,2023-03-06 09:14:09,2023-03-07 23:30:03,CCBill,-19.95,0.00,CHF,COMPLETED,113.18
CARD_PAYMENT,Current,2023-03-08 03:06:33,2023-03-08 18:34:22,GitHub,-7.54,0.00,CHF,COMPLETED,105.64
CARD_PAYMENT,Current,2023-03-12 13:48:50,,Contabo,-6.34,0.06,CHF,REVERTED,
CARD_PAYMENT,Current,2023-03-18 09:05:30,2023-03-21 11:32:07,Contabo,-6.44,0.00,CHF,COMPLETED,99.20
TRANSFER,Current,2023-03-25 20:17:05,2023-03-25 20:17:10,To GARY PETER BYRNE,-30.00,0.00,CHF,COMPLETED,69.20
TOPUP,Current,2023-03-25 20:17:05,2023-03-25 20:17:15,Auto Top-Up by *0382,75.00,0.00,CHF,COMPLETED,144.20
CARD_PAYMENT,Current,2023-03-22 13:52:30,2023-03-28 20:28:02,Contabo,-11.25,0.00,CHF,COMPLETED,132.95
CARD_PAYMENT,Current,2023-04-01 19:41:19,2023-04-02 10:48:11,Google Cloud,-12.50,0.00,CHF,COMPLETED,120.45
CARD_PAYMENT,Current,2023-04-05 09:13:54,2023-04-06 23:16:17,CCBill,-19.88,0.00,CHF,COMPLETED,100.57
CARD_PAYMENT,Current,2023-04-12 12:45:25,2023-04-18 21:18:30,Contabo,-6.37,0.00,CHF,COMPLETED,94.20
CARD_PAYMENT,Current,2023-05-05 09:13:52,2023-05-07 00:58:36,CCBill,-19.54,0.00,CHF,COMPLETED,74.66
CARD_PAYMENT,Current,2023-06-04 09:13:44,2023-06-06 01:37:35,CCBill,-19.50,0.20,CHF,COMPLETED,54.96
CARD_PAYMENT,Current,2023-06-12 22:52:27,,Migros,-1.90,0.00,CHF,REVERTED,
CARD_PAYMENT,Current,2023-07-08 08:06:21,2023-07-08 20:53:22,Sky,-14.90,0.00,CHF,COMPLETED,40.06
CARD_PAYMENT,Current,2023-08-01 19:23:16,2023-08-02 10:43:45,Google Cloud,-7.00,0.00,CHF,COMPLETED,33.06
EXCHANGE,Current,2023-08-22 20:58:34,2023-08-22 20:58:34,Exchanged to USD,-8.80,0.00,CHF,COMPLETED,24.26
CARD_PAYMENT,Current,2023-09-01 21:12:54,2023-09-02 11:19:08,Google Cloud,-7.75,0.00,CHF,COMPLETED,16.51
CARD_PAYMENT,Current,2023-09-11 23:01:47,2023-09-12 13:55:10,Gst,-7.45,0.00,CHF,COMPLETED,9.06
CARD_PAYMENT,Current,2023-09-12 00:57:32,2023-09-12 21:52:25,Confiteria Antojos,-4.94,0.00,CHF,COMPLETED,4.12
CARD_PAYMENT,Current,2023-10-23 17:56:02,2023-10-24 13:33:24,Coop,-1.45,0.00,CHF,COMPLETED,2.67
"#;
        insta::assert_debug_snapshot!(Revolut::parse("Neon", TRANSACTIONS.to_string()).unwrap());
    }
}
