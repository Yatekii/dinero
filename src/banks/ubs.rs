use std::io::Cursor;

use polars::{
    datatypes::DataType,
    io::SerReader,
    lazy::{
        dsl::{col, lit},
        frame::{IntoLazy, LazyFrame},
    },
    prelude::*,
};

use super::Parser;

pub struct Ubs {}

impl Parser for Ubs {
    fn parse(content: String) -> anyhow::Result<LazyFrame> {
        let df = CsvReadOptions::default()
            .with_parse_options(
                CsvParseOptions::default()
                    .with_separator(b';')
                    .with_try_parse_dates(true)
                    .with_truncate_ragged_lines(true),
            )
            .with_has_header(true)
            .into_reader_with_file_handle(Cursor::new(&content))
            .finish()?;

        #[allow(clippy::possible_missing_comma)]
        let first = df
            .clone()
            .lazy()
            .select([
                col("Balance").alias("Credit").last()
                    - col("Debit")
                        .fill_null(col("Credit"))
                        .str()
                        .replace(lit("'"), lit(""), true)
                        .cast(DataType::Float64)
                        .last(),
                lit(NULL).cast(DataType::Float64).alias("Debit"),
                col("Booking date").last(),
                col("Description1").last(),
            ])
            .collect()
            .unwrap();

        let df = concat([first.lazy(), df.lazy().reverse()], UnionArgs::default())
            .unwrap()
            .select(&[
                col("Booking date").alias("Date"),
                col("Debit")
                    .fill_null(col("Credit"))
                    .alias("Amount")
                    .str()
                    .replace(lit("'"), lit(""), true)
                    .cast(DataType::Float64),
                col("Description1").alias("Description"),
            ])
            .with_column(lit("").alias("Category"));

        Ok(df)
    }
}

#[cfg(test)]
mod tests {
    use crate::banks::Parser;

    const TRANSACTIONS: &str = r#"Trade date;Trade time;Booking date;Value date;Currency;Debit;Credit;Individual amount;Balance;Transaction no.;Description1;Description2;Description3;Footnotes;
2021-01-29;;2021-01-29;2021-01-31;CHF;-10.00;;;22113.15;BL01529HJ0125142;"Balance closing of service prices";;"Transaction no. BL01529HJ0125142";;
2021-01-27;;2021-01-28;2021-01-28;CHF;-503.50;;;22123.15;9906527KH9626550;"UBS Switzerland AG,c/o UBS Card Center";"VIS1W WIDERSPRUCH AN UBS INNERT 30 TAGEN, direct debit";"Reference no. 70 03130 00000 00326 20405 08188, Account no. IBAN: CH25 0023 0230 0129 5305 U, Costs: LSV direct debit, Transaction no. 9906527KH9626550";;
2021-01-27;;2021-01-27;2021-01-27;CHF;-62.75;;;22626.65;9906026TO5927425;"CAISSE DES MEDECINS,1211 GENEVE";"e-banking payment order";"Reference no. 05 62945 01846 09123 50099 10841, Account no. 01-037005-8, Costs: E-Banking domestic, Transaction no. 9906026TO5927425";;
2021-01-22;;2021-01-22;2021-01-22;CHF;;9.59;;22689.40;9999022ZC8003633;"STRIPE PAYMENTS UK LTD,9TH FLOOR, 107 CHEAPSIDE GB - LONDO, N EC2V 6DN";"credit";"Reason for payment: GITHUB SPONSORS R8N0A2, Costs: Incoming SIC-payment, Transaction no. 9999022ZC8003633";;
2021-01-22;;2021-01-22;2021-01-22;CHF;;3688.00;;22679.81;9999022ZC7962684;"TECHNOKRAT GMBH,UNTERROHRSTRASSE 5, 8952 SCHLIEREN, CH";"credit";"Reason for payment: LOHN, Costs: Incoming SIC-payment, Transaction no. 9999022ZC7962684";;
2021-01-18;;2021-01-18;2021-01-18;CHF;-482.80;;;18991.81;BF21018DJ2743561;"CSS KRANKEN-VERSICHERUNG,AG, LUZERN";"EBILL-RECHNUNG, PayNet Order";"Reference no. 00 00002 64029 44001 71051 49861, Account no. 01-070393-3, Costs: E-Banking domestic, Transaction no. BF21018DJ2743561";;
2021-01-13;;2021-01-13;2021-01-13;CHF;-820.00;;;19474.61;9906012TI0715972;"Stichting DEGIRO,NL";"e-banking payment order";"Reason for payment: 88391-243702-5FFD9A1F-8BB5, Account no. IBAN: CH82 0871 0039 1145 1200 2, Costs: E-Banking domestic, Transaction no. 9906012TI0715972";;
2021-01-13;;2021-01-13;2021-01-13;CHF;-20.00;;;20294.61;9906013GK1225794;"ROTH, JONAS, Debit UBS TWINT";;"Reason for payment: +41798181674, TWINT-Acc.:+41799607130, Transaction no. 9906013GK1225794";;
2021-01-04;;2021-01-04;2021-01-04;CHF;-350.00;;;20314.61;9906501KH9243834;"GENERALI PERSONENVERSICHERUNGEN AG,8134 ADLISWIL";"FORT1 WIDERSPRUCH AN UBS INNERT 30 TAGEN, direct debit";"Reference no. 91 34065 09797 16210 10157 11525, Costs: LSV direct debit, Transaction no. 9906501KH9243834";;
2021-01-04;;2021-01-04;2021-01-04;CHF;-281.35;;;20664.61;BH21004DJ2411724;"CSS KRANKEN-VERSICHERUNG,AG, LUZERN";"EBILL-RECHNUNG, PayNet Order";"Reference no. 00 00002 64029 44002 90851 13989, Account no. 01-070393-3, Costs: E-Banking domestic, Transaction no. BH21004DJ2411724";;"#;

    #[test]
    fn parse() {
        let _df = super::Ubs::parse(TRANSACTIONS.into()).unwrap();
    }
}
