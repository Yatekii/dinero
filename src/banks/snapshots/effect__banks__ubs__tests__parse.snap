---
source: src/banks/ubs.rs
expression: "super::Ubs::parse(\"UBS\".to_string(), TRANSACTIONS.into()).unwrap()"
---
ParsedAccount {
    ledgers: [
        ParsedLedger {
            name: "UBS",
            records: [
                LedgerRecord {
                    date: 2021-01-29,
                    amount: -10.0,
                    description: "Balance closing of service prices",
                    category: "",
                    symbol: "CHF",
                },
                LedgerRecord {
                    date: 2021-01-28,
                    amount: -503.5,
                    description: "UBS Switzerland AG,c/o UBS Card Center",
                    category: "",
                    symbol: "CHF",
                },
                LedgerRecord {
                    date: 2021-01-27,
                    amount: -62.75,
                    description: "CAISSE DES MEDECINS,1211 GENEVE",
                    category: "",
                    symbol: "CHF",
                },
                LedgerRecord {
                    date: 2021-01-22,
                    amount: 9.59,
                    description: "STRIPE PAYMENTS UK LTD,9TH FLOOR, 107 CHEAPSIDE GB - LONDO, N EC2V 6DN",
                    category: "",
                    symbol: "CHF",
                },
                LedgerRecord {
                    date: 2021-01-22,
                    amount: 3688.0,
                    description: "TECHNOKRAT GMBH,UNTERROHRSTRASSE 5, 8952 SCHLIEREN, CH",
                    category: "",
                    symbol: "CHF",
                },
                LedgerRecord {
                    date: 2021-01-18,
                    amount: -482.8,
                    description: "CSS KRANKEN-VERSICHERUNG,AG, LUZERN",
                    category: "",
                    symbol: "CHF",
                },
                LedgerRecord {
                    date: 2021-01-13,
                    amount: -820.0,
                    description: "Stichting DEGIRO,NL",
                    category: "",
                    symbol: "CHF",
                },
                LedgerRecord {
                    date: 2021-01-13,
                    amount: -20.0,
                    description: "ROTH, JONAS, Debit UBS TWINT",
                    category: "",
                    symbol: "CHF",
                },
                LedgerRecord {
                    date: 2021-01-04,
                    amount: -350.0,
                    description: "GENERALI PERSONENVERSICHERUNGEN AG,8134 ADLISWIL",
                    category: "",
                    symbol: "CHF",
                },
                LedgerRecord {
                    date: 2021-01-04,
                    amount: -281.35,
                    description: "CSS KRANKEN-VERSICHERUNG,AG, LUZERN",
                    category: "",
                    symbol: "CHF",
                },
            ],
        },
    ],
}
